packages/client/src/generation/utils.ts

```diff
@@ -344,3 +349,73 @@ export function unique<T>(arr: T[]): T[] {
 
   return result
 }
+
+/**
+ * Determines if arg types need generic <$PrismaModel> argument added.
+ * Essentially, performs breadth-first search for any fieldRefTypes that
+ * do not have corresponding `meta.source` defined.
+ *
+ * @param type
+ * @returns
+ */
+export function needsGenericModelArg(type: DMMF.SchemaArgInputType) {
+  const toVisit = [type]
+  const visited = new Set<string>()
+
+  let current: DMMF.SchemaArgInputType | undefined
+  while ((current = toVisit.shift())) {
+    const name = getTypeName(current)
+    if (visited.has(name)) {
+      // if the type is recursive, let other fields make a decision
+      continue
+    }
+    visited.add(name)
+
+    if (current.location === 'fieldRefTypes') {
+      return true
+    }
+
+    if (current.location === 'inputObjectTypes' && typeof current.type === 'object') {
+      const inputType = current.type as DMMF.InputType
+      if (!inputType.fields) {
+        continue
+      }
+
+      if (inputType.meta?.source) {
+        // if source is defined, we know model for sure and do not need generic argument
+        continue
+      }
+
+      for (const field of inputType.fields) {
+        toVisit.push(...field.inputTypes)
```

Comment:
Interesting way of achieving recursion with a loop. This might not be obvious at first, would you mind adding a comment?

Reply:
Added a comment at the start of the function